### PR TITLE
Shor's algorithm, the classical parts

### PR DESCRIPTION
**Pull request by viathor**
This implements the classical probabilistic reduction of factoring to order-finding.

Example use:

```
$ python examples/shor.py $((907 * 911))
907 is a non-trivial factor of 826277
$
```

Follow-up: Shor's algorithm, the quantum parts.

### CODE REVIEW

File: examples/shor.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+[1]: https://arxiv.org/abs/quant-ph/9508027
+"""
+
+import argparse
+import math
+import random
+
+from typing import Callable, Optional
+
+parser = argparse.ArgumentParser('Factorization demo.')
```
[the remark this answers is not in the document]
Comment (viathor) [reply examples/shor.py]: Oops, done.

File: examples/shor.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+k is the number of distinct prime factors of n. Since the case k = 1 has been
+handled by the classical part, we have k >= 2 and the success probability of
+a single attempt is at least 1/2.
+
+The subroutine for finding the order r of a number x modulo n consists of two
+steps. In the first step, Quantum Phase Estimation is applied to a unitary such
+as
+
+    U|y) = |xy mod n)  0 <= y < n
+    U|y) = |y)         n =< y
```
[the remark this answers is not in the document]
Comment (viathor) [reply examples/shor.py]: Cool! Thanks!

File: examples/shor.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    Returns:
+        Non-trivial factor of n. Factor k of n is trivial if it is 1 or n.
+        Loops forever when n is prime.
+    """
+    if n % 2 == 0:
+        return 2
+    c = find_factor_of_prime_power(n)
+    if c is not None:
+        return c
+    while True:
```
[the remark this answers is not in the document]
Comment (viathor) [reply examples/shor.py]: Yup. I considered adding a primality test, but decided it's better not to obscure the algorithm. I've added a `max_attempts` parameter to `find_factor()`.

File: examples/shor.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            return c  # coverage: ignore
+        r = order_finder(x, n)
+        if r % 2 != 0:
+            continue  # coverage: ignore
+        y = x**(r // 2) % n
+        c = math.gcd(y - 1, n)
+        if 1 < c < n:
+            return c
+        # coverage: ignore
+        c = math.gcd(y + 1, n)
```
[the remark this answers is not in the document]
Comment (viathor) [reply examples/shor.py]: Good catch! Done.

File: examples/shor.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+    Returns:
+        Non-trivial factor of n or None if no such factor was found.
+        Factor k of n is trivial if it is 1 or n.
+    """
+    if n == 2:
+        return None
+    if n % 2 == 0:
+        return 2
+    c = find_factor_of_prime_power(n)
```
Comment (Strilanc) [inline examples/shor.py]: Use `sympy.primetest` https://docs.sympy.org/latest/_modules/sympy/ntheory/primetest.html to avoid the infinite loop.

Comment (viathor) [reply examples/shor.py]: Done.

**Pull request merged**